Add Square Clustering (first draft)

eriknw opened the pull request:
I'm sure this is a little rough around the edges.  I wanted to see what's it's like adding a new algorithm to LAGraph.

This computes square clustering:
https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cluster.square_clustering.html

It's based on the Python implementation here:
https://github.com/python-graphblas/graphblas-algorithms/blob/8de0be644df7daeb4d0b031db285705659a57e53/graphblas_algorithms/algorithms/cluster.py#L287-L323

Going from this Python to C code was super-easy.  I just copy/pasted the output from our C call recorder and reformatted a little:
https://github.com/python-graphblas/graphblas-algorithms/pull/14#issuecomment-1151418484

The hardest part was writing a test.  I used our recorder to help write C here too.  It should have more tests, such as against invalid input.

Counting the number of squares is from https://arxiv.org/pdf/2007.11111.pdf, which is part of experimental/algorithm/LAGraph_FastGraphletTransform.c.

Determining all the factors for the denominator was actually kind of tricky for me.  We use equation 2 from this paper for the denominator:
https://arxiv.org/pdf/0710.0117v1.pdf

To my knowledge, square clustering hasn't been implemented in GraphBLAS before.

This is untested for directed graphs and graphs with self-edges.  It is also not benchmarked or optimized against benchmarking.  Since we assume symmetric structure, we have some freedom when to transpose some inputs.

This uses a few GxB operators:
- GxB_PLUS_PAIR_INT64 (is there a GrB equivalent for this yet?)
- GxB_ANY_TIMES_INT64 (monoid doesn't matter, so could be replaced)
- GxB_RMINUS_INT64 (by using RMINUS, we avoid creating another vector)

Happy to accept any feedback.  Also fine if this sits here a while (since, you know, we're supposed to be writing documentation and getting ready for 1.0 instead of writing new algorithms!).

eriknw committed:
Add Square Clustering (first draft; a little rough around the edges)

eriknw commented:
Also, I had to run cmake with this:
```bash
cmake .. -DCMAKE_C_FLAGS="-lrt"
```
Otherwise, the clock timer stuff doesn't compile for me.  Has anybody else encountered this?

eriknw commented:
@szarnyasg any thoughts on test failures?

szarnyasg commented:
@eriknw the CI configuration had an error in the `if` condition, so it tried to deploy from other branches as well. It should be fixed -- could you please try rebasing on `reorg` to include the latest commit?

eriknw committed:
Merge branch 'reorg' into square_clustering

szarnyasg commented:
Okay, this is better but `ctest_SquareClustering` does not pass with SuiteSparse:GraphBLAS 6.x.
Do we want to keep testing in CI with 6.x, @DrTimothyAldenDavis?

eriknw committed:
Fix GrB_Matrix_diag with older SuiteSparse:GraphBLAS

eriknw commented:
Thanks @szarnyasg.  All green!

DrTimothyAldenDavis commented:
We don't need to keep the tests working for SuiteSparse:GraphBLAS v6.x.  That version had the mis-specification for GrB_Matrix_diag, and we really should just stick with v7.x as a result.

DrTimothyAldenDavis commented:
For these GxB objects:

GxB_PLUS_PAIR_INT64 (is there a GrB equivalent for this yet?):  there is no GrB_PLUS_ONEB_INT64 but there is an LAGraph_plus_one_int64.  The latter is technically user-defined while GxB_PLUS_PAIR_INT64 is built-in, but they are identical in terms of performance.  They use the same underlying code inside SuiteSparse:GrB.

GxB_ANY_TIMES_INT64 (monoid doesn't matter, so could be replaced):  If the monoid isn't used, you could just use GrB_PLUS_TIMES_INT64, and it would be the same.  Is this a diagonal scaling, like C=A*D?  In that case, I have specialized row/col scaling methods.  They ignore the monoid entirely.

GxB_RMINUS_INT64 (by using RMINUS, we avoid creating another vector):  there is no GrB equivalent.  You could either create the extra vector, or use a user-defined RMINUS.  Creating the extra vector would be faster.  I would do this in an #ifdef so SS:GrB can use RMINUS and be faster.

Regarding the -lrt library linking: I haven't seen that, probably because I always compile LAGraph and GraphBLAS with OpenMP, and the LAGraph_WallClockTime (formerly LAGraph_Tic/Toc) uses omp_get_wtime(), which doesn't need -lrt.  See this comment:

src/utility/LAGraph_WallClockTime.c
52:        // resolution clock_gettime instead.  May require -lrt

DrTimothyAldenDavis reviewed (approved):
Looks great!

eriknw committed:
Pure GrB version

eriknw commented:
Thanks for the review @DrTimothyAldenDavis!

I have updated this to have a pure GrB version.  I knew I've seen something like `LAGraph_plus_one_int64` before.

> Is this a diagonal scaling, like C=A*D? In that case, I have specialized row/col scaling methods. They ignore the monoid entirely.

Yes, this does scaling with a diagonal matrix D.  Is your specialized methods in LAGraph or SuiteSparse:GraphBLAS?  Does it make a difference whether D is transposed or not?

> I would do this in an #ifdef so SS:GrB can use RMINUS and be faster.

`#if LAGRAPH_SUITESPARSE` added.  Using RMINUS lets us do it in one command instead of four (counting variable declaration as one).

> Regarding the -lrt library linking: I haven't seen that, probably because I always compile LAGraph and GraphBLAS with OpenMP, and the LAGraph_WallClockTime (formerly LAGraph_Tic/Toc) uses omp_get_wtime(), which doesn't need -lrt.

Hmm, I compile these with OpenMP too (at least I believe I do).  Strange.  It would be nice if cmake were smart enough to add `-lrt` automatically if necessary; it's always frustrating to users/developers when compilation doesn't "just work".  I can create instructions for creating an environment where `-ltr` is necessary to add.

eriknw committed:
Oops.  Make sure we free all objects.

eriknw commented:
Thanks for reviewing.  I'll try to add a better discussion at the beginning sometime in the next few days.

szarnyasg commented:
The build is green now (after force-restarting it once). Maybe add the summary in the comment and then this seems good to be merged.

eriknw committed:
Faster: rearrange terms to avoid expensive GrB_Matrix_eWiseMult_BinaryOp

Believe it or not, `P2 * (P2 - 1)` is actually a dominating term when I profile this,
so, yes, it's worthwhile to re-use P2 to calculate the above as `P2 *= P2 - 1`.

Also, free large objects ASAP, but peak memory usage is unchanged.

DrTimothyAldenDavis reviewed (approved):
Looks great!
